File: Control/xAODRootAccess/python/TPyStore.py
# Copyright (C) 2002-2020 CERN for the benefit of the ATLAS collaboration
#
# Module holding the TPyStore Python class
#

# Pull in ROOT:
import ROOT

## A Python wrapper around xAOD::TPyStore
#
# In order to make the semi-templated functions of xAOD::TPyStore more
# convenient to use from PyROOT, the user is supposed to use an instance
# of this class, and not ROOT.xAOD.TPyStore directly.
#
# @author Attila Krasznahorkay <Attila.Krasznahorkay@cern.ch>
#
class TPyStore( ROOT.xAOD.TPyStore ):

    ## Constructor for the class
    def __init__( self ):

        # Forward the call to the base class's constructor:
        ROOT.xAOD.TPyStore.__init__( self )
        return

    ## Convenient version of the base class's contains function
    #
    # This function allows the user to, instead of figuring out the exact
    # type name of some C++ type, to rather write code like:
    #
    # <code>
    #   if store.contains( "Electrons", ROOT.xAOD.ElectronContainer_v1 ):
    # </code>
    #
    # @param key  The string key of the object to check for
    # @param type The type of the object we are looking for
    # @returns <code>True</code> if the object is available in the store,
    #          <code>False</code> if it's not
    #
    def contains( self, key, type ):
        # Determine the class name:
        clname = type.__name__
        if hasattr( type, "__cpp_name__" ):
            clname = type.__cpp_name__
            pass
        # Call the parent class's function:
        return super( TPyStore, self ).contains( key, clname )

    ## Convenient version of the base class's isConst function
    #
    # This function allows the user to, instead of figuring out the exact
    # type name of some C++ type, to rather write code like:
    #
    # <code>
    #   if store.isConst( "Electrons", ROOT.xAOD.ElectronContainer_v1 ):
    # </code>
    #
    # @param key  The string key of the object to check for
    # @param type The type of the object we are looking for
    # @returns <code>True</code> if the object is only available as a constant,
    #          <code>False</code> if it's not
    #
    def isConst( self, key, type ):
        # Determine the class name:
        clname = type.__name__
        if hasattr( type, "__cpp_name__" ):
            clname = type.__cpp_name__
            pass
        # Call the parent class's function:
        return super( TPyStore, self ).isConst( key, clname )

    ## Convenient version of the base class's record function
    #
    # This is the main function of this wrapper class. It allows the user to
    # record objects into the transient store like:
    #
    # <code>
    #   el = ROOT.xAOD.ElectronContainer_v1()
    #   store.record( el, "MyElectrons" )
    # </code>
    #
    # @param obj The object to be recorded into the output file
    # @param key The key (branch name) for the object
    # @returns <code>xAOD::TReturnCode::kSuccess</code> if all was successful,
    #          or <code>xAOD::TReturnCode::kFailure</code> if not
    #
    def record( self, obj, key ):
        # Determine the class name:
        clname = obj.__class__.__name__
        if hasattr( obj.__class__, "__cpp_name__" ):
            clname = obj.__class__.__cpp_name__
            pass
        # Call the parent class's function:
        print("Recording",key,clname)
        return super( TPyStore, self ).record( obj, key, clname )

File: Control/xAODRootAccess/Root/TPyStore.cxx
/*
  Copyright (C) 2002-2020 CERN for the benefit of the ATLAS collaboration
*/

// ROOT include(s):
#include <TClass.h>
#include <TError.h>

// Local include(s):
#include "xAODRootAccess/TPyStore.h"
#include "xAODRootAccess/tools/ReturnCheck.h"

namespace xAOD {

   /// This function can be used in the same manner as
   /// TStore::contains<...>(...), but instead of providing a type, one gives it
   /// a type name.
   ///
   /// Note that only types that have a proper ROOT dictionary can be used.
   /// Unlike C++, which allows one to insert any type of object into TStore.
   ///
   /// @param key  The key of the object that we're looking for
   /// @param type The type name of the object we're trying to access
   /// @returns <code>kTRUE</code> if the object is acessible,
   ///          <code>kFALSE</code> otherwise
   ///
   ::Bool_t TPyStore::contains( const std::string& key,
                                const std::string& type ) const {

      // Try to access the dictionary of this type:
      ::TClass* cl = ::TClass::GetClass( type.c_str() );
      if( ! cl ) {
         ::Warning( "xAOD::TPyStore::contains",
                    "Type name \"%s\" not known", type.c_str() );
         return kFALSE;
      }

      // Check if the dictionary can return a type_info:
      const std::type_info* ti = cl->GetTypeInfo();
      if( ! ti ) {
         ::Warning( "xAOD::TPyStore::contains",
                    "Type \"%s\" doesn't have a proper dictionary",
                    type.c_str() );
         return kFALSE;
      }

      // Use the base class to answer the question:
      return TStore::contains( key, *ti );
   }

   /// This function can be used in the same manner as
   /// TStore::isConst<...>(...), but instead of providing a type, one gives it
   /// a type name.
   ///
   /// Note that only types that have a proper ROOT dictionary can be used.
   /// Unlike C++, which allows one to insert any type of object into TStore.
   ///
   /// @param key  The key of the object that we're looking for
   /// @param type The type name of the object we're trying to access
   /// @returns <code>kTRUE</code> if the object is acessible,
   ///          <code>kFALSE</code> otherwise
   ///
   ::Bool_t TPyStore::isConst( const std::string& key,
                               const std::string& type ) const {

      // Try to access the dictionary of this type:
      ::TClass* cl = ::TClass::GetClass( type.c_str() );
      if( ! cl ) {
         ::Warning( "xAOD::TPyStore::isConst",
                    "Type name \"%s\" not known", type.c_str() );
         return kFALSE;
      }

      // Check if the dictionary can return a type_info:
      const std::type_info* ti = cl->GetTypeInfo();
      if( ! ti ) {
         ::Warning( "xAOD::TPyStore::isConst",
                    "Type \"%s\" doesn't have a proper dictionary",
                    type.c_str() );
         return kFALSE;
      }

      // Use the base class to answer the question:
      return TStore::isConst( key, *ti );
   }

   /// This function can be used in the same manner as
   /// TStore::record<...>(...), but instead of providing a type, one gives it
   /// a type name and a typeless pointer.
   ///
   /// Note that only types that have a proper ROOT dictionary can be used.
   /// Unlike C++, which allows one to insert any type of object into TStore.
   ///
   /// Also note that this function doesn't take ownership of the recorded
   /// object. In Python all the objects created by the interpreter are managed
   /// by the interpreter. So this code is not supposed to delete them,
   /// otherwise all hell breaks loose.
   ///
   /// @param obj  Pointer to the object to be put into the store
   /// @param key  Key of the object in the store
   /// @param type The type name of the object we are inserting
   /// @returns The usual xAOD::TReturnCode types
   ///
   TReturnCode TPyStore::record( void* obj, const std::string& key,
                                 const std::string& type ) {

      // Simply forward the call to the appropriate function from the base
      // class:
      RETURN_CHECK( "xAOD::TPyStore::record",
                    TStore::record( obj, key, type, kFALSE ) );

      // Return gracefully:
      return TReturnCode::kSuccess;
   }

   /// This is just a convenience function, to make it easier to print the
   /// contents of such objects in Python. Since the base class's print()
   /// function has a special meaning in Python.
   ///
   void TPyStore::dump() const {

      print();
      return;
   }

} // namespace xAOD
